src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorAbsSendTime.java

```diff
@@ -0,0 +1,627 @@
+            {
+                long sendDeltaMs = probe.sendTimeMs - prevSendTime;
+                long recvDeltaMs = probe.recvTimeMs - prevRecvTime;
+
+                if(sendDeltaMs >= 1 && recvDeltaMs >= 1)
+                {
+                    ++current.numAboveMinDelta;
+                }
+
+                if(!IsWithinClusterBounds(sendDeltaMs,current))
+                {
+                    if(current.count >= kMinClusterSize)
+                    {
+                        clusters = addCluster(clusters,current);
+                    }
+                    current = new Cluster();
+                }
+                current.sendMeanMs += sendDeltaMs;
+                current.recvMeanMs += recvDeltaMs;
+                current.meanSize += probe.payloadSize;
+                ++current.count;
+            }
+
+            prevSendTime = probe.sendTimeMs;
+            prevRecvTime = probe.recvTimeMs;
+        }
+        if(current.count >= kMinClusterSize)
+            clusters = addCluster(clusters, current);
+        return clusters;
+    }
+
+    /**
+     * @param clusters
+     * @returns a cluster that shows the best probe
+     */
+    Cluster findBestProbe(List<Cluster> clusters)
+    {
+        int highestProbeBitrateBps = 0;
+        Cluster bestIt = new Cluster();
+        for (Cluster cluster : clusters) {
+            if (cluster.sendMeanMs == 0 || cluster.recvMeanMs == 0)
+                continue;
+            if (cluster.numAboveMinDelta > cluster.count / 2 &&
+                    (cluster.recvMeanMs - cluster.sendMeanMs <= 2.0f &&
+                            cluster.sendMeanMs - cluster.recvMeanMs <= 5.0f))
+            {
+                int probeBitrateBps =
+                        Math.min(cluster.getSendBitrateBps(), cluster
+                                .getRecvBitrateBps());
+                if (probeBitrateBps > highestProbeBitrateBps)
+                {
+                    highestProbeBitrateBps = probeBitrateBps;
+                    bestIt = cluster;
+                }
+            }
+            else
+            {
+                /**
+                 * @Todo
+                 * What do we loose by converting from double to long ?
+                 */
+                double sendBitrateBps = cluster.meanSize * 8 * 1000
+                        / cluster.sendMeanMs;
+                double recvBitrateBps = cluster.meanSize * 8 * 1000
+                        / cluster.recvMeanMs;
+                logger.warn( "Probe failed, sent at " + sendBitrateBps
+                        + " bps, received at " + recvBitrateBps
+                        + " bps. Mean send delta: " + cluster.sendMeanMs
+                        + " ms, mean recv delta: " + cluster.recvMeanMs
+                        + " ms, num probes: " + cluster.count);
+                break;
+            }
+        }
+        return bestIt;
+    }
+
+    probeResult processClusters(long nowMs)
+    {
+        synchronized (critSect) {
```

Comment:
I think method declaration of processClusters (in line 109 of [webrtc abs_send_time.h]
 (https://chromium.googlesource.com/external/webrtc/+/master/webrtc/modules/remote_bitrate_estimator/remote_bitrate_estimator_abs_send_time.h))uses locks.